### PR TITLE
[kwok] Provided service for exec

### LINKED ISSUE #236: [kwok] Provided service for `exec`
``` yaml
# Match pods
kind: ClusterExec
apiVersion: kwok.x-k8s.io/v1alpha1
metadata:
  name: cluster-exec-rules
spec:
  selector:
    matchNamespaces:
      - podNamespace
    matchNames:
      - podName
  execs: []
--- 
# Just for a Pod
kind: Exec
apiVersion: kwok.x-k8s.io/v1alpha1
metadata:
  name: podName
  namespace: podNamespace
spec:
  execs:
    - containers:
        - containerName
      local:
        workDir: "/workdir"
        envs:
          - <envkey>=<envvalue>

      # Maybe support ssh in future
      # ssh:
      #   address:
      #  # todo
```

/kind feature

### PR DESCRIPTION
**Pull request by wzshiming**
#### What type of PR is this?

/kind featureg

#### Which issue(s) this PR fixes:

Fixes #236

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by wzshiming** [comment]
Hi @nikola-jokic, would you be interested in reviewing `exec`, I've implemented it and only have tests left. currently struggling with the `--stdin` and `--tty` parameters, It seems that the test cannot cover this.

File: pkg/apis/internalversion/cluster_exec_types.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ClusterExec provides cluster-wide exec configuration.
+type ClusterExec struct {
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta
+	// Spec holds spec for cluster port forward.
```
Comment (nikola-jokic) [inline pkg/apis/internalversion/cluster_exec_types.go:28]:
```suggestion
	// Spec holds spec for cluster exec.
```

File: pkg/apis/v1alpha1/cluster_exec_types.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterExec provides cluster-wide exec configuration.
+type ClusterExec struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata"`
+	// Spec holds spec for cluster port forward.
```
Comment (nikola-jokic) [inline pkg/apis/v1alpha1/cluster_exec_types.go:36]:
```suggestion
	// Spec holds spec for cluster exec.
```

File: pkg/apis/v1alpha1/cluster_exec_types.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	Spec ClusterExecSpec `json:"spec"`
+}
+
+// ClusterExecSpec holds spec for cluster port forward.
```
Comment (nikola-jokic) [inline pkg/apis/v1alpha1/cluster_exec_types.go:40]:
```suggestion
// ClusterExecSpec holds spec for cluster exec.
```

File: pkg/kwok/server/debugging_exec.go [hunk 4]
```diff
@@ -21,23 +21,108 @@ import (
...
+func (s *Server) getExecTarget(podName, podNamespace string, containerName string) (*internalversion.ExecTarget, error) {
+	pf, has := slices.Find(s.config.Execs, func(pf *internalversion.Exec) bool {
+		return pf.Name == podName && pf.Namespace == podNamespace
+	})
+	if has {
+		exec, found := findContainerInExecs(containerName, pf.Spec.Execs)
+		if found {
+			return exec, nil
+		}
+	} else {
```
Comment (nikola-jokic) [inline pkg/kwok/server/debugging_exec.go:96]: Question: If we find exec that is misconfigured (for example, pod name and pod namespace are matching, but there are no containers named `containerName`, should we fall back to the cluster execs or return not found?
This is also important for the logging PR :smile: 

If we should return an error, I'd suggest return early and remove else
Then the error message can also specify where did we stop execution so it is more understandable to users where to look for.
For example, saying that pod matching exec spec does not have a container with container name can be a signal that the `Exec` configuration is wrong

Comment (wzshiming) [reply pkg/kwok/server/debugging_exec.go]: makes sense 😄

File: pkg/kwok/server/debugging_exec.go [hunk 5]
```diff
@@ -21,23 +21,108 @@ import (
...
+		}
+		return nil, fmt.Errorf("not found exec target for container %s in pod %s/%s", containerName, podNamespace, podName)
+	}
+
+	for _, ce := range s.config.ClusterExecs {
+		if !ce.Spec.Selector.Match(podName, podNamespace) {
+			continue
+		}
+
+		forward, found := findContainerInExecs(containerName, ce.Spec.Execs)
```
Comment (nikola-jokic) [inline pkg/kwok/server/debugging_exec.go:104]: Sorry, again NIT: forward should be e or exec :relaxed:

Comment (wzshiming) [reply pkg/kwok/server/debugging_exec.go]: My mistake, I missed it again. done😄

File: pkg/kwok/server/debugging_exec.go [hunk 6]
```diff
@@ -21,23 +21,108 @@ import (
...
+func (s *Server) getExecTarget(podName, podNamespace string, containerName string) (*internalversion.ExecTarget, error) {
+	pf, has := slices.Find(s.config.Execs, func(pf *internalversion.Exec) bool {
+		return pf.Name == podName && pf.Namespace == podNamespace
+	})
+	if has {
+		exec, found := findContainerInExecs(containerName, pf.Spec.Execs)
+		if found {
+			return exec, nil
+		}
+		return nil, fmt.Errorf("not found exec target for container %s in pod %s/%s", containerName, podNamespace, podName)
```
Comment (nikola-jokic) [inline pkg/kwok/server/debugging_exec.go:96]: Super NIT, please ignore me if you think this is not useful

How about using `%q` for container?
The message would be like:
> not found exec target for container "fake" in pod fake-pod/default`

Comment (nikola-jokic) [inline pkg/kwok/server/debugging_exec.go:96]:
```suggestion
		return nil, fmt.Errorf("exec target not found for container %s in pod %s/%s", containerName, podNamespace, podName)
```
This suggestion is without %q
%q is just a question, not that important

File: pkg/kwok/server/debugging_exec.go [hunk 7]
```diff
@@ -21,23 +21,108 @@ import (
...
+		if !ce.Spec.Selector.Match(podName, podNamespace) {
+			continue
+		}
+
+		exec, found := findContainerInExecs(containerName, ce.Spec.Execs)
+		if found {
+			return exec, nil
+		}
+	}
+	return nil, fmt.Errorf("no exec found for pod %s/%s container %s", podNamespace, podName, containerName)
```
Comment (nikola-jokic) [inline pkg/kwok/server/debugging_exec.go:109]:
```suggestion
	return nil, fmt.Errorf("no exec found for container %s in pod %s/%s", containerName, podNamespace, podName)
```

**Comment by wzshiming** [comment]
Thank you for your review it was very helpful 😄

**Comment by nikola-jokic** [comment]
You are welcome! Thank you for asking me :smile:

**Pull request merged**